src/commands/execution_ctx.c

```diff
@@ -0,0 +1,85 @@
+/*
+* Copyright 2018-2020 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include "execution_ctx.h"
+#include "../query_ctx.h"
+
+// Invalid single instance or execution context.
+static ExecutionCtx invalid_ctx = {0};
+
+static ExecutionType _GetExecutionTypeFromAST(AST *ast) {
+	const cypher_astnode_type_t root_type = cypher_astnode_type(ast->root);
+	if(root_type == CYPHER_AST_QUERY) return EXECUTION_TYPE_QUERY;
+	if(root_type == CYPHER_AST_CREATE_NODE_PROPS_INDEX) return EXECUTION_TYPE_INDEX_CREATE;
+	if(root_type == CYPHER_AST_DROP_NODE_PROPS_INDEX) return EXECUTION_TYPE_INDEX_DROP;
+	assert(false && "Uknown execution type");
+}
+
+void ExecutionInformation_FromQuery(const char *query, ExecutionPlan **plan, AST **ast,
+									ExecutionType *exec_type, bool *cached) {
```

Comment:
I feel like this function is overloaded and should probably be split into more discrete calls.

It currently:
1) Parses parameters
2) Attempts cache retrieval
3) Parses query, builds AST
4) Builds entire ExecutionPlan
5) Updates the cache

Reply:
This flow is the reason I wrote this function. I don't think cmd_xxxxx should handle those tasks. We should have a designated objects/functions for this, and less objects and functions to handle in the cmd_xxxx handlers is better IMO.